### DISCUSSION TITLE
How to improve the myloader import performance of 'restoring constraints'

### DISCUSSION THREAD

**Discussion by YuanQY** [mydumper/mydumper#1182] [Q&A] [answered]
Please find the corrected text below:

I used the following command line to restore about 180GB of data to an AWS RDS MySQL database:

`
myloader -h "$database" -u "$username" -p "$password" -B "$database_name" -t 64 -d "$directory_name" --innodb-optimize-keys AFTER_IMPORT_ALL_TABLES --queries-per-transaction=500 --max-threads-per-table=16 --serialized-table-creation --compress-protocol
`

I observed that myloader restored the data in 30 minutes, but the restoration of constraints took about 1 hour and 30 minutes. Although myloader used multiple threads for restoring data and indexes, it used only one thread for restoring constraints of tables. During the restoration of constraints, the MySQL IO and CPU usage were very low. I would like to know how to increase the number of threads used by myloader for restoring constraints.

MySQL CPU
<img width="868" alt="image" src="https://github.com/mydumper/mydumper/assets/1910991/80ed1a9e-963b-4874-8cd6-611fd0953898">

MySQL IO
<img width="880" alt="image" src="https://github.com/mydumper/mydumper/assets/1910991/6b62c6a7-8e88-44b1-b4d8-843f2bada255">

**Comment by davidducos** [comment]
Mmmm strange, can you share the log? You should use -v 3 and -L file.log
And what's the version of myloader and database?

**Reply by YuanQY** [reply]
@davidducos 
Please check the log. The myloader version is v0.14.2-10. The MySQL version is 5.7.41. The log includes export.log(mydumper) and import.log(myloader)
[dump_load.log.tgz](https://github.com/mydumper/mydumper/files/11878514/dump_load.log.tgz)

**Reply by davidducos** [reply]
Hi @YuanQY,
You are using an old prerelease, why don't you try with the latest prerelease?
The export.log is useless. In the import.log are errors because you are not using --skip-definer.
About the issue, I think that the FOREIGN_KEY_CHECKS is not being disable and that is why it is using COPY:
```
The INPLACE algorithm is supported when [foreign_key_checks](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_foreign_key_checks) is disabled. Otherwise, only the COPY algorithm is supported.
```
Reference: https://dev.mysql.com/doc/refman/5.7/en/innodb-online-ddl-operations.html#online-ddl-foreign-key-operations

**Reply by YuanQY** [reply]
Hi @davidducos 

    I tested myloader version v0.15.0-2 and used 'CALL mysql.rds_set_fk_checks_off();' to disable foreign key checks before restoring the data. It can now use multiple threads to restore constraints, but there is a strange problem. I found that thread 54 takes a long time in single-threaded mode. After approximately 35 minutes, it starts the multiple-threaded mode and other threads begin restoring constraints. Please check the log for more information. I think there may be a potential problem.

[import.log.gz](https://github.com/mydumper/mydumper/files/11889529/import.log.gz)

Thread for restore constraints statistics (I use command ' grep " restoring constraints" import.log | cut -d ":" -f3 | cut -d "-" -f2 | sort | uniq -c')
      1  Thread 1
      1  Thread 10
      1  Thread 11
      1  Thread 12
      1  Thread 13
      1  Thread 14
      1  Thread 15
      1  Thread 16
      1  Thread 17
      1  Thread 18
      1  Thread 19
      1  Thread 2
      1  Thread 20
      1  Thread 21
      1  Thread 22
      1  Thread 23
      1  Thread 24
      8  Thread 25
      1  Thread 26
      1  Thread 27
      1  Thread 28
      1  Thread 29
      1  Thread 3
      1  Thread 30
      1  Thread 31
      1  Thread 32
      1  Thread 33
      1  Thread 34
      1  Thread 35
      1  Thread 36
      1  Thread 37
      1  Thread 38
      1  Thread 39
      1  Thread 4
     11  Thread 40
      1  Thread 41
      1  Thread 42
      1  Thread 43
      1  Thread 44
      1  Thread 45
     47  Thread 46
      1  Thread 47
      1  Thread 48
      1  Thread 49
      1  Thread 5
      1  Thread 50
      1  Thread 51
      1  Thread 52
      1  Thread 53
     318  Thread 54
     56  Thread 55
      1  Thread 56
      1  Thread 57
      1  Thread 58
      1  Thread 59
      1  Thread 6
      1  Thread 60
      1  Thread 61
      1  Thread 62
      1  Thread 63
      1  Thread 64
      1  Thread 7
     45  Thread 8
      1  Thread 9

IOPS
![image](https://github.com/mydumper/mydumper/assets/1910991/a0375ace-d83d-41cf-80d4-0b727e7e63c2)

CPU
![image](https://github.com/mydumper/mydumper/assets/1910991/ff97edf9-154d-4498-a009-b76b46c71b26)

**Reply by davidducos** [reply]
Hi @YuanQY,
If you go to the first ERROR in the import.log, you will find:
```
2023-06-28 02:44:20 [INFO] - restoring index: PerfTest.sales_targets
2023-06-28 02:44:20 [INFO] - Thread 3: restoring indexes `PerfTest`.`sales_targets` from index. Tables 0 of 565 completed
2023-06-28 02:48:34 [WARNING] - Thread 3: Error restoring 1: Lost connection to MySQL server during query
2023-06-28 02:48:34 [WARNING] - Thread 8: Error restoring 1: Lost connection to MySQL server during query
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION WAIT_TIMEOUT = 2147483
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION WAIT_TIMEOUT = 2147483
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION NET_WRITE_TIMEOUT = 2147483
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION NET_WRITE_TIMEOUT = 2147483
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION AUTOCOMMIT = 0
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION AUTOCOMMIT = 0
2023-06-28 02:48:34 [WARNING] - Set session failed: /*!40101 SET SESSION SQL_MODE = 'NO_AUTO_VALUE_ON_ZERO'  */
2023-06-28 02:48:34 [WARNING] - Set session failed: /*!40101 SET SESSION SQL_MODE = 'NO_AUTO_VALUE_ON_ZERO'  */
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION SQL_LOG_BIN = 0
2023-06-28 02:48:34 [WARNING] - Set session failed: SET SESSION SQL_LOG_BIN = 0
2023-06-28 02:48:34 [WARNING] - Set session failed: /*!40114 SET SESSION FOREIGN_KEY_CHECKS = 0  */
2023-06-28 02:48:34 [WARNING] - Set session failed: /*!40114 SET SESSION FOREIGN_KEY_CHECKS = 0  */
2023-06-28 02:48:34 [WARNING] - Set session failed: /*!40114 SET SESSION UNIQUE_CHECKS = 0  */
2023-06-28 02:48:34 [WARNING] - Thread 3: Retrying last failed executed statement
2023-06-28 02:48:34 [WARNING] - Set session failed: /*!40114 SET SESSION UNIQUE_CHECKS = 0  */
2023-06-28 02:48:34 [ERROR] - Thread 3: Error restoring: Can't connect to MySQL server on 'qingyun-sync.cmeep8hwwlid.rds.cn-north-1.amazonaws.com.cn' (111)
2023-06-28 02:48:34 [INFO] - Failed indexes: ALTER TABLE `PerfTest`.`sales_targets`
 ADD  KEY `sales_targets_hospital_id_60db07bb_fk_accounts_id` (`hospital_id`),
 ADD  KEY `sales_targets_package_id_728dbf60_fk_packages_id` (`package_id`),
 ADD  KEY `sales_targets_position_id_062fbb5d_fk_positions_id` (`position_id`),
 ADD  KEY `sales_targets_product_id_28b60b9a_fk_products_id` (`product_id`),
 ADD  KEY `sales_targets_9b8089a8` (`yearmonth`),
 ADD  KEY `sales_targets_department_id_ef9f69e7_uniq` (`department_id`),
 ADD  KEY `sales_targets_department_category_id_6df77cd9_uniq` (`department_category_id`),
 ADD  KEY `sales_targets_created_by_b870c568_fk_auth_user_id` (`created_by`),
 ADD  KEY `sales_targets_modified_by_22b7774a_fk_auth_user_id` (`modified_by`),
 ADD  KEY `sales_targets_bba4d9cf` (`geography_id`),
 ADD  KEY `sales_targets_262c9190` (`global_positioning_system_id`),
 ADD  KEY `sales_targets_72122ce9` (`owner`),
 ADD  KEY `sales_targets_6c31fb9d` (`record_type_id`),
 ADD  KEY `sales_targets_d5582625` (`state_id`),
 ADD  KEY `sales_targets_e8701ad4` (`user_id`),
 ADD  KEY `sales_targets_c4b59e37` (`package_price_id`),
 ADD  KEY `sales_targets_71fa9940` (`position_history_id`),
 ADD  KEY `sales_targets_e18299ed` (`position_path`),
 ADD  KEY `sales_targets_46985d74` (`edit_type_id`),
 ADD  KEY `idx_year_group` (`position_path`,`hospital_id`,`product_id`,`package_id`,`user_id`,`department_id`),
 ADD  KEY `idx_sum_filter` (`state_id`,`yearmonth`,`quantity`),
 ADD  KEY `sales_targets_external_id_86ee3006_uniq` (`external_id`),
 ADD  KEY `idx_year_group_2` (`position_id`,`hospital_id`,`department_id`,`package_id`),
 ADD  KEY `sales_targets_custom_lookup_10_id_509297da` (`custom_lookup_10_id`),
 ADD  KEY `sales_targets_custom_lookup_1_id_d2daf8e6` (`custom_lookup_1_id`),
 ADD  KEY `sales_targets_custom_lookup_2_id_a77e3a4d` (`custom_lookup_2_id`),
 ADD  KEY `sales_targets_custom_lookup_3_id_2b57ba4f` (`custom_lookup_3_id`),
 ADD  KEY `sales_targets_custom_lookup_4_id_7c31e1ce` (`custom_lookup_4_id`),
 ADD  KEY `sales_targets_custom_lookup_5_id_7266e537` (`custom_lookup_5_id`),
 ADD  KEY `sales_targets_custom_lookup_6_id_b137d038` (`custom_lookup_6_id`),
 ADD  KEY `sales_targets_custom_lookup_7_id_0f7cecbb` (`custom_lookup_7_id`),
 ADD  KEY `sales_targets_custom_lookup_8_id_cf5de65d` (`custom_lookup_8_id`),
 ADD  KEY `sales_targets_custom_lookup_9_id_7ae37e8f` (`custom_lookup_9_id`);

2023-06-28 02:48:34 [WARNING] - Thread 8: Retrying last failed executed statement
2023-06-28 02:48:34 [ERROR] - Thread 8: Error restoring: Can't connect to MySQL server on 'qingyun-sync.cmeep8hwwlid.rds.cn-north-1.amazonaws.com.cn' (111)
```
The server tried for 4 minutes to create that huge amount of indexes and failed. I think that the server crashed. You might need to set --max-threads-for-index-creation to 2 to reduce the amount of threads that are creating indexes.

**Reply by YuanQY** [reply]
Hi @davidducos,

Please check the latest log. I did not set the parameter '--max-threads-for-index-creation' to reduce the number of threads for creating indexes, and there are no errors in the log. However, I noticed that 'thread 1' is processing the majority of the restoring constraints tasks. It same as not all threads work in concurrency at the first 15 minutes of restoring constraints. What's your opinion?

    343  Thread 1
      1  Thread 10
      1  Thread 11
      1  Thread 12
      1  Thread 13
      1  Thread 14
      1  Thread 15
      1  Thread 16
      1  Thread 17
      1  Thread 18
      1  Thread 19
      1  Thread 2
      1  Thread 20
      1  Thread 21
      1  Thread 22
      1  Thread 23
      1  Thread 24
      1  Thread 25
      1  Thread 26
      1  Thread 27
      1  Thread 28
      1  Thread 29
      1  Thread 3
      1  Thread 30
      1  Thread 31
      1  Thread 32
      1  Thread 33
      1  Thread 34
      1  Thread 35
      1  Thread 36
      1  Thread 37
      1  Thread 38
      1  Thread 39
      1  Thread 4
      1  Thread 40
      1  Thread 41
    138  Thread 42
      1  Thread 43
      1  Thread 44
      1  Thread 45
      1  Thread 46
      1  Thread 47
      1  Thread 48
      1  Thread 49
      1  Thread 5
      1  Thread 50
      1  Thread 51
      1  Thread 52
      1  Thread 53
      1  Thread 54
      1  Thread 55
      1  Thread 56
      1  Thread 57
      1  Thread 58
      1  Thread 59
      1  Thread 6
      1  Thread 60
      1  Thread 61
      1  Thread 62
      1  Thread 63
      1  Thread 64
      1  Thread 7
      1  Thread 8
      1  Thread 9

 
[import.log.zip](https://github.com/mydumper/mydumper/files/11901198/import.log.zip)

IOPS
<img width="860" alt="image" src="https://github.com/mydumper/mydumper/assets/1910991/e7077b42-cfc1-409e-8704-84cf848d850b">

CPU
<img width="856" alt="image" src="https://github.com/mydumper/mydumper/assets/1910991/c6bae086-552a-4cc9-804b-2921663f3c25">

**Comment by davidducos** [comment]
Can you test with https://github.com/mydumper/mydumper/releases/tag/v0.15.0-7 ?

**Comment by davidducos** [comment]
Can you test with https://github.com/mydumper/mydumper/releases/tag/v0.15.0-7 ?

**Reply by YuanQY** [reply]
Hi @davidducos 
    I tested the same case in version 0.15.0-7. The test result is similar to the previous version. There is only one thread restoring constraints most of the time. Please check the log; there are no errors.

[import.log.gz](https://github.com/mydumper/mydumper/files/11912740/import.log.gz)

Database CPU
![image](https://github.com/mydumper/mydumper/assets/1910991/631c50d9-7acb-402f-a784-650773277af9)

Database Write IOPS
![image](https://github.com/mydumper/mydumper/assets/1910991/a35d41b0-373b-45fc-bbea-eedfb791fa7b)

Thread for restore constraints statistics
    130  Thread 1
      1  Thread 10
      1  Thread 11
      1  Thread 12
      1  Thread 13
      1  Thread 14
      1  Thread 15
    162  Thread 16
      1  Thread 17
      1  Thread 18
      1  Thread 19
      1  Thread 2
      1  Thread 20
      1  Thread 21
      1  Thread 22
      1  Thread 23
      1  Thread 24
      1  Thread 25
      1  Thread 26
      1  Thread 27
      1  Thread 28
      1  Thread 29
      1  Thread 3
      1  Thread 30
      1  Thread 31
      1  Thread 32
     55  Thread 33
    133  Thread 34
      1  Thread 35
      1  Thread 36
      4  Thread 37
      1  Thread 38
      1  Thread 39
      1  Thread 4
      1  Thread 40
      1  Thread 41
      1  Thread 42
      1  Thread 43
      1  Thread 44
      1  Thread 45
      1  Thread 46
      1  Thread 47
      1  Thread 48
      1  Thread 49
      1  Thread 5
      1  Thread 50
      1  Thread 51
      1  Thread 52
      1  Thread 53
      1  Thread 54
      1  Thread 55
      1  Thread 56
      1  Thread 57
      1  Thread 58
      1  Thread 59
      1  Thread 6
      1  Thread 60
      1  Thread 61
      1  Thread 62
      1  Thread 63
      1  Thread 64
      1  Thread 7
      1  Thread 8
      1  Thread 9

**Comment by davidducos** [comment]
Hi @YuanQY, in the code there is mutex to force to create one constraint at a time https://github.com/mydumper/mydumper/blob/5d5490e0d8833d822030f75107c43742aa2aa537/src/myloader_worker_loader.c#L125
I created https://github.com/mydumper/mydumper/pull/1191 to see if there are issues doing it with multiple threads.

**Reply by davidducos** [reply]
@YuanQY is it possible for you to test that version? as I don't want to merge an untested feature request, as I think that you will have contention and threads are going fail to create constraints in a multiple thread fashion. Please, let me know if you know how to compile that branch or if you need a package for it.

**Reply by YuanQY** [reply]
Hi @davidducos 
    I have tested the changes by merging the updated code into version 0.15.0-7, and it has passed my tests (There is no error in my test case). Myloader can now work in a multi-threaded model. But it doesn't improve any performance.

[import.log.20230703.tgz](https://github.com/mydumper/mydumper/files/11935369/import.log.20230703.tgz)

IOPS
<img width="822" alt="image" src="https://github.com/mydumper/mydumper/assets/1910991/9c1c8a8b-c365-4f33-a8f2-4aaf675e11a3">

CPU 

 
<img width="730" alt="image" src="https://github.com/mydumper/mydumper/assets/1910991/a5849452-5003-4882-a99e-78262a187078">

Thread for restore constraints statistics

      4  Thread 1
     13  Thread 10
     12  Thread 11
     14  Thread 12
      4  Thread 13
      2  Thread 14
      8  Thread 15
      6  Thread 16
     11  Thread 17
     15  Thread 18
      1  Thread 19
      3  Thread 2
      5  Thread 20
      4  Thread 21
      8  Thread 22
     13  Thread 23
      8  Thread 24
      5  Thread 25
      5  Thread 26
     10  Thread 27
     11  Thread 28
      7  Thread 29
      8  Thread 3
      7  Thread 30
      7  Thread 31
      4  Thread 32
     12  Thread 33
      8  Thread 34
      5  Thread 35
     10  Thread 36
     12  Thread 37
      8  Thread 38
      7  Thread 39
     13  Thread 4
      7  Thread 40
     15  Thread 41
     11  Thread 42
     13  Thread 43
      6  Thread 44
      4  Thread 45
      8  Thread 46
     16  Thread 47
      8  Thread 48
      9  Thread 49
     12  Thread 5
      9  Thread 50
     11  Thread 51
      5  Thread 52
     14  Thread 53
      6  Thread 54
      9  Thread 55
      6  Thread 56
      5  Thread 57
      3  Thread 58
     15  Thread 59
      6  Thread 6
      1  Thread 60
      7  Thread 61
      8  Thread 62
     10  Thread 63
     14  Thread 64
     11  Thread 7
     12  Thread 8
     12  Thread 9

**Reply by davidducos** [reply, marked as answer]
@YuanQY, nice that the timings had improved, but I'm not confortable with this solution as in the past I implemented this kind of solution and then we had to rolled it back. So, I will be adding an option to use multiple threads to add constraints or implement in the same way that we did with --max-threads-for-schema-creation.
The main reason of stopping it, is related to the changes in MySQL 8 that limits the concurrency on the schema modifications that, for instance, ended up with this PR https://github.com/mydumper/mydumper/pull/554.
